instructure/canvas-ios
# Courses Smart Search Bar

refs: MBL-17907,MBL-17906,MBL-17905,MBL-17904,MBL-17903,MBL-17902
affects: Student
release note: Courses Smart Search Experience for Student

Related PR(s):
https://github.com/instructure/canvas-ios/pull/2967

## Changes Overview

### Presentation Methodology:
By configuring `CoreHostingController.navigationItem` with button and textfield, the button is used to show/hide textfield. When submitting text field, another `CoreHostingController` with search content view is presented in full screen. Search content includes views to show states of loading, failure, no-results and results. 

Thus, we have two main places in code where search bar is configured: 
- `CoreSearchHostingController` navigationItem. (see method `setupOrRemoveSearchBar()`)
   And this is to be the main entry point for the search experience.
- Search content `SearchContentContainerView` toolbar.

### Experience Setup
In order to customize the search experience appearance and content per the presented context, the following protocols was introduced:

- `SearchViewAttributes`: Those information that are required to fully configure the appearance of search textfield and show/hide button.

Those information would also be passed to inner views, starting from the root one (`SearchContentContainerView`).
The technique used for that was to store it to environment object of `SearchViewContext<>` type. 

It worths noting that the implementation can include much more information than those that are required by this protocol, it depends on the needs of your search experience inner views. 

- `SearchViewsProvider`: Used to configure SwiftUI views for search content, filter editor and support sheet. The benefit of this type is basically to encapsulate the presentation logic of showing/configuring these within search experience.

As, a side note, to configure help button on the search bar, you need to return an instance of `SearchSupportButtonModel` in this protocol implementation for `support` property. Passing one of the predefined actions (`SearchSupportClosureAction`, `SearchSupportSheetAction`)  to `action` argument of the initializer. If you chose to return `nil` for that `support` property, help button won't be shown. 

- `SearchInteractor`: This is to be used only in `CoreSearchHostingController`, as this the entry point for the experience. And it would be used to find out about the visibility of search button on the navigation bar. The `isEnabled` property, is to be implemented in any suitable way to find out whether to make the search experience available or not, such as fetching data through API or local database. 

![Untitled](https://github.com/user-attachments/assets/2a763e16-22d3-4883-a4dc-6f66ba1a672f)

### Search View Context

A concrete class type `SearchViewContext<>`  was introduced to encapsulate key variables and publishers that need to be accessed or listened to across any part of the experience view, and at any time through during the experience presentation. Like holding a history of visited row IDs which are used to distinguish visited results. 

## Test Plan
* Make sure to activate "Smart Search" on account & course level. [See here.](https://community.canvaslms.com/t5/Smart-Search-Feature-Preview/How-to-Enable-Smart-Search/ta-p/604577)
* Login to Student app.
* Go to courses, open one of the courses. 
* You should expect a search icon shown on the top-right button. Follow on the experience.
   - Showing search results
   - Selecting Filter
   - Showing grouped results
   - Highlight visited results

## Video Record

https://github.com/user-attachments/assets/71107a56-c18a-47c8-8a75-3927a7424d61

## Checklist
